pkg/objstore/oss/oss.go

```diff
@@ -0,0 +1,365 @@
+package oss
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"math/rand"
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	alioss "github.com/aliyun/aliyun-oss-go-sdk/oss"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+	yaml "gopkg.in/yaml.v2"
+)
+
+type Config struct {
+	Endpoint        string `yaml:"endpoint"`
+	Bucket          string `yaml:"bucket"`
+	AccessKeyID     string `yaml:"access_key_id"`
+	AccessKeySecret string `yaml:"access_key_secret"`
+}
+
+type Bucket struct {
+	name   string
+	logger log.Logger
+	client *alioss.Client
+	config Config
+	bucket *alioss.Bucket
+}
+
+func NewTestBucket(t testing.TB) (objstore.Bucket, func(), error) {
+	c := configFromEnv()
+	err := func(conf Config) error {
+		if conf.Endpoint == "" {
+			return errors.New("no aliyun oss endpoint in config file")
+		}
+		if conf.AccessKeyID == "" {
+			return errors.New("access_key_id is not present in config file")
+		}
+		if conf.AccessKeySecret == "" {
+			return errors.New("access_key_secret is not present in config file")
+		}
+		return nil
+	}(c)
+	if err != nil {
+		return nil, nil, err
+	}
+	if c.Bucket != "" && os.Getenv("THANOS_ALLOW_EXISTING_BUCKET_USE") == "true" {
+		t.Log("ALIYUNOSS_BUCKET is defined. Normally this tests will create temporary bucket " +
+			"and delete it after test. Unset ALIYUNOSS_BUCKET env variable to use default logic. If you really want to run " +
+			"tests against provided (NOT USED!) bucket, set THANOS_ALLOW_EXISTING_BUCKET_USE=true.")
+		return NewTestBucketFromConfig(t, c, true)
+	}
+	return NewTestBucketFromConfig(t, c, false)
+}
+
+func calculateChunks(name string, r io.Reader) (int, int64, error) {
+	switch r.(type) {
+	case *os.File:
+		f, _ := r.(*os.File)
+		if fileInfo, err := f.Stat(); err == nil {
+			s := fileInfo.Size()
+			return int(math.Floor(float64(s) / alioss.MaxPartSize)), s % alioss.MaxPartSize, nil
+		}
```

Comment:
I think we should upload smaller part each time, not 5GB each part, things will fail.

Reply:
I agree with you ,  the smaller part uploading will accelerate compacting . But  there is a another problem ,  see `UploadPart` method in multipart.go (https://github.com/aliyun/aliyun-oss-go-sdk/blob/master/oss/multipart.go)   , you can not  excess ten thousand times for the multi-part upload task . This means , if  I upload every part exactly 100M , I can upload to Aliyun OSS 1TiB object at most .As compacting works on and on , I think it may be not enough .

Reply:
I think you can define part size like below:
partSize = max(100MB, fileSize / 10001)